R/zzz.R

```diff
@@ -80,19 +80,19 @@ check_time <- function(time_ranges) {
 
 get_widget <- function(comparison_item, category, gprop, hl, cookie_url, tz) {
   token_payload <- list()
+  #comparison_item$keyword <- gsub(pattern="\\+",replacement="%2B",comparison_item$keyword)
   token_payload$comparisonItem <- comparison_item
   token_payload$category <- category
   token_payload$property <- gprop
 
   # token_payload$comparisonItem$keyword <- curl::curl_escape(token_payload$comparisonItem$keyword)
   
-  url <- URLencode(paste0(
-    "https://www.google.com/trends/api/explore?property=&req=",
-    jsonlite::toJSON(token_payload, auto_unbox = TRUE),
-    "&tz=",tz,"&hl=", hl
-  ))
+  url <- paste0(URLencode("https://www.google.com/trends/api/explore?property=&req="),
+                URLencode(paste0(jsonlite::toJSON(token_payload, auto_unbox = TRUE)),reserved=T),
+                URLencode(paste0("&tz=",tz,"&hl=", hl))
+  )
   
-  url <- encode_keyword(url)
+  #url <- encode_keyword(url)
```

Comment:
Please do not comment-out randomly chosen lines of code.  Revert this, and open a ticket to discuss, please.

Reply:
@eddelbuettel  It's not randomly commented. The code for encode_keyword reads:

```
encode_keyword <- function(url) {
  url <- gsub("(?:\\G(?!^)|\\[\\s*)[^][\\s]*\\K\\&(?!])(?=[^][]*])", "%26", url, perl = TRUE)
  url <- gsub("(?:\\G(?!^)|\\[\\s*)[^][\\s]*\\K\\&(?!])(?=[^][]*])", "%26", url, perl = TRUE)
  url <- gsub("(?:\\G(?!^)|\\[\\s*)[^][\\s]*\\K\\&(?!])(?=[^][]*])", "%26", url, perl = TRUE)
  url <- gsub("(?:\\G(?!^)|\\[\\s*)[^][\\s]*\\K\\&(?!])(?=[^][]*])", "%26", url, perl = TRUE)
  gsub("(?:\\G(?!^)|\\[\\s*)[^][\\s]*\\K\\&(?!])(?=[^][]*])", "%26", url, perl = TRUE)
}
```
It replaces all ampersands with %26. This is already done in code lines 90 and 91... 

So why would one want to encode the keywords after having used the URL-encoding for reserved symbols? This is useless here in my opinion.

Best, JB

Reply:
It's about process. I am not in this code day-to-day, but (mostly) Philippe (and to a much lesser extend I) have our named on it.  You come out of left field and uncomment code currently on CRAN.  I just don't operate that way.

There way be merit to the change. If so, great, because that should make it easy for you to motivate 
a) why, 
b) what and (very important) 
c) absence of side effects on existing functionality.

With that I hand it back to @PMassicotte.

Reply:
Exactly. That's why I am asking: Why is it in there? a) I think it is useless it replaces amerpsands, b) I think my code works, it also replaces ampersands in the keywords without complicated regex to circumvent the ampersands that need to be there for the paramters... c) I haven't encountered one.

But a) b) c) may be answered by somebody more involved in the project than I am very differently. I just ran across this since I wanted to include the "logical operator functionality"... so please, who inserted it and why?

best, jb

Reply:
What is standard procedure for your point c)?

Reply:
A poor (but feasible) proxy for c) is unit tests; a more thorough one we do with important packages is reverse dependency testing.   

Here only failing queries could tell us and so far we don't know as we see none failing.  Also, as you point out, the existing code with the four chained regexps is not elegant but it is there for a reason.  You could use `git blame` to find the commit it came from and maybe find a test case or two from the issues around it.

Reply:
I will have a look at `git blame` and do some code archeology ;D Thx I really didn't know about this...

Reply:
So, after having used git blame the changes of url_encode go back to Sven Thies and Victor Yammouni. The comment says 
`# Replace special characters in keywords like P&500 -> P%26500 `
So to have this functionality included in the current code, I have used in the function URLencode the option reserved=TRUE for all requests to Google APIS in the code, the part in which the keyword is handed to the API. This encodes the special characters. The functionality is fulfilled. 
Test 
`gtrends("S&P500",time="all")`